diff --git a/lib/Saml2/AuthnRequest.php b/lib/Saml2/AuthnRequest.php
index e6e49a2..b5414bb 100644
--- a/lib/Saml2/AuthnRequest.php
+++ b/lib/Saml2/AuthnRequest.php
@@ -122,10 +122,15 @@ REQUESTEDAUTHN;
     Version="2.0"
 {$providerNameStr}{$forceAuthnStr}{$isPassiveStr}
     IssueInstant="$issueInstant"
-    Destination="{$idpData['singleSignOnService']['url']}"
+    Destination="{$idpData['entityId']}"
     ProtocolBinding="{$spData['assertionConsumerService']['binding']}"
-    AssertionConsumerServiceURL="{$spData['assertionConsumerService']['url']}">
-    <saml:Issuer>{$spData['entityId']}</saml:Issuer>
+    AssertionConsumerServiceURL="{$spData['assertionConsumerService']['url']}"
+    AttributeConsumingServiceIndex="0">
+    <saml:Issuer
+        NameQualifier="{$spData['entityId']}"
+        Format="urn:oasis:names:tc:SAML:2.0:nameid-format:entity">
+        {$spData['entityId']}
+    </saml:Issuer>
 {$nameIdPolicyStr}
 {$requestedAuthnStr}
 </samlp:AuthnRequest>
diff --git a/lib/Saml2/Metadata.php b/lib/Saml2/Metadata.php
index 5662624..1f9582e 100644
--- a/lib/Saml2/Metadata.php
+++ b/lib/Saml2/Metadata.php
@@ -142,7 +142,7 @@ ATTRIBUTEVALUE;
 
             $requestedAttributeStr = implode(PHP_EOL, $requestedAttributeData);
             $strAttributeConsumingService = <<<METADATA_TEMPLATE
-<md:AttributeConsumingService index="1">
+<md:AttributeConsumingService index="0">
             <md:ServiceName xml:lang="en">{$sp['attributeConsumingService']['serviceName']}</md:ServiceName>
 {$attrCsDesc}{$requestedAttributeStr}
         </md:AttributeConsumingService>
@@ -152,14 +152,13 @@ METADATA_TEMPLATE;
         $metadata = <<<METADATA_TEMPLATE
 <?xml version="1.0"?>
 <md:EntityDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata"
-                     validUntil="{$validUntilTime}"
-                     cacheDuration="PT{$cacheDuration}S"
                      entityID="{$sp['entityId']}">
     <md:SPSSODescriptor AuthnRequestsSigned="{$strAuthnsign}" WantAssertionsSigned="{$strWsign}" protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol">
 {$sls}        <md:NameIDFormat>{$sp['NameIDFormat']}</md:NameIDFormat>
         <md:AssertionConsumerService Binding="{$sp['assertionConsumerService']['binding']}"
                                      Location="{$sp['assertionConsumerService']['url']}"
-                                     index="1" />
+                                     isDefault="true"
+                                     index="0" />
         {$strAttributeConsumingService}
     </md:SPSSODescriptor>{$strOrganization}{$strContacts}
 </md:EntityDescriptor>
diff --git a/lib/Saml2/Response.php b/lib/Saml2/Response.php
index 6c43068..d6c19af 100644
--- a/lib/Saml2/Response.php
+++ b/lib/Saml2/Response.php
@@ -268,10 +268,11 @@ class OneLogin_Saml2_Response
                 foreach ($issuers as $issuer) {
                     $trimmedIssuer = trim($issuer);
                     if (empty($trimmedIssuer) || $trimmedIssuer !== $idPEntityId) {
-                        throw new OneLogin_Saml2_ValidationError(
-                            "Invalid issuer in the Assertion/Response (expected '$idPEntityId', got '$trimmedIssuer')",
-                            OneLogin_Saml2_ValidationError::WRONG_ISSUER
-                        );
+                        // See https://github.com/italia/spid-regole-tecniche/issues/20
+                        // throw new OneLogin_Saml2_ValidationError(
+                        //     "Invalid issuer in the Assertion/Response (expected '$idPEntityId', got '$trimmedIssuer')",
+                        //     OneLogin_Saml2_ValidationError::WRONG_ISSUER
+                        // );
                     }
                 }
