diff --git a/lib/Saml2/AuthnRequest.php b/lib/Saml2/AuthnRequest.php
index e6e49a2..be566fa 100644
--- a/lib/Saml2/AuthnRequest.php
+++ b/lib/Saml2/AuthnRequest.php
@@ -122,10 +122,15 @@ REQUESTEDAUTHN;
     Version="2.0"
 {$providerNameStr}{$forceAuthnStr}{$isPassiveStr}
     IssueInstant="$issueInstant"
-    Destination="{$idpData['singleSignOnService']['url']}"
+    Destination="{$idpData['entityId']}"
     ProtocolBinding="{$spData['assertionConsumerService']['binding']}"
-    AssertionConsumerServiceURL="{$spData['assertionConsumerService']['url']}">
-    <saml:Issuer>{$spData['entityId']}</saml:Issuer>
+    AssertionConsumerServiceURL="{$spData['assertionConsumerService']['url']}"
+    AttributeConsumingServiceIndex="1">
+    <saml:Issuer
+        NameQualifier="{$spData['entityId']}"
+        Format="urn:oasis:names:tc:SAML:2.0:nameid-format:entity">
+        {$spData['entityId']}
+    </saml:Issuer>
 {$nameIdPolicyStr}
 {$requestedAuthnStr}
 </samlp:AuthnRequest>
diff --git a/lib/Saml2/LogoutRequest.php b/lib/Saml2/LogoutRequest.php
index 79c04de..33e9a8d 100644
--- a/lib/Saml2/LogoutRequest.php
+++ b/lib/Saml2/LogoutRequest.php
@@ -92,7 +92,11 @@ class OneLogin_Saml2_LogoutRequest
     Version="2.0"
     IssueInstant="{$issueInstant}"
     Destination="{$idpData['singleLogoutService']['url']}">
-    <saml:Issuer>{$spData['entityId']}</saml:Issuer>
+    <saml:Issuer
+        NameQualifier="{$spData['entityId']}"
+        Format="urn:oasis:names:tc:SAML:2.0:nameid-format:entity">
+        {$spData['entityId']}
+    </saml:Issuer>
     {$nameIdObj}
     {$sessionIndexStr}
 </samlp:LogoutRequest>
diff --git a/lib/Saml2/Metadata.php b/lib/Saml2/Metadata.php
index 5662624..2ef0fd8 100644
--- a/lib/Saml2/Metadata.php
+++ b/lib/Saml2/Metadata.php
@@ -152,13 +152,12 @@ METADATA_TEMPLATE;
         $metadata = <<<METADATA_TEMPLATE
 <?xml version="1.0"?>
 <md:EntityDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata"
-                     validUntil="{$validUntilTime}"
-                     cacheDuration="PT{$cacheDuration}S"
                      entityID="{$sp['entityId']}">
     <md:SPSSODescriptor AuthnRequestsSigned="{$strAuthnsign}" WantAssertionsSigned="{$strWsign}" protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol">
 {$sls}        <md:NameIDFormat>{$sp['NameIDFormat']}</md:NameIDFormat>
         <md:AssertionConsumerService Binding="{$sp['assertionConsumerService']['binding']}"
                                      Location="{$sp['assertionConsumerService']['url']}"
+                                     isDefault="true"
                                      index="1" />
         {$strAttributeConsumingService}
     </md:SPSSODescriptor>{$strOrganization}{$strContacts}
diff --git a/lib/Saml2/Response.php b/lib/Saml2/Response.php
index 6c43068..d6c19af 100644
--- a/lib/Saml2/Response.php
+++ b/lib/Saml2/Response.php
@@ -268,10 +268,11 @@ class OneLogin_Saml2_Response
                 foreach ($issuers as $issuer) {
                     $trimmedIssuer = trim($issuer);
                     if (empty($trimmedIssuer) || $trimmedIssuer !== $idPEntityId) {
-                        throw new OneLogin_Saml2_ValidationError(
-                            "Invalid issuer in the Assertion/Response (expected '$idPEntityId', got '$trimmedIssuer')",
-                            OneLogin_Saml2_ValidationError::WRONG_ISSUER
-                        );
+                        // See https://github.com/italia/spid-regole-tecniche/issues/20
+                        // throw new OneLogin_Saml2_ValidationError(
+                        //     "Invalid issuer in the Assertion/Response (expected '$idPEntityId', got '$trimmedIssuer')",
+                        //     OneLogin_Saml2_ValidationError::WRONG_ISSUER
+                        // );
                     }
                 }
